// @onchain-id/identity-sdk/dist/core/utils/Utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultAbiCoder = exports.arrayify = exports.encodeAndHash = exports.isValidURI = exports.isHexString = exports.toHex = void 0;
const utils_1 = require("ethers/lib/utils");
Object.defineProperty(exports, "arrayify", { enumerable: true, get: function () { return utils_1.arrayify; } });
Object.defineProperty(exports, "defaultAbiCoder", { enumerable: true, get: function () { return utils_1.defaultAbiCoder; } });
function toHex(message) {
    return (0, utils_1.hexlify)((0, utils_1.toUtf8Bytes)(message));
}
exports.toHex = toHex;
function isHexString(value) {
    return typeof value === 'string' && (0, utils_1.isHexString)(value);
}
exports.isHexString = isHexString;
/**
 * Check if an URI is valid.
 * @param value
 * @return True if the uri is secured.
 */
function isValidURI(value) {
    return typeof value === 'string' && value.match(/^https?:\/\/.*\..*/) !== null;
}
exports.isValidURI = isValidURI;
function encodeAndHash(types, values) {
    return (0, utils_1.keccak256)(utils_1.defaultAbiCoder.encode(types, values));
}
exports.encodeAndHash = encodeAndHash;
