// @onchain-id/identity-sdk/dist/identity-sdk.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.constants = exports.utils = exports.SignerModule = exports.Claim = exports.Identity = void 0;
const Identity_1 = require("./identity/Identity");
Object.defineProperty(exports, "Identity", { enumerable: true, get: function () { return Identity_1.Identity; } });
const Claim_1 = require("./claim/Claim");
Object.defineProperty(exports, "Claim", { enumerable: true, get: function () { return Claim_1.Claim; } });
const SignerModule_1 = require("./core/SignerModule");
Object.defineProperty(exports, "SignerModule", { enumerable: true, get: function () { return SignerModule_1.SignerModule; } });
const utils = __importStar(require("./utils-module"));
exports.utils = utils;
const constants = {
    implementationAuthorities: {
        ropsten: '0x95c0242daa0b3113a3d9cA45b4E6421a10DB77b6',
        kovan: '0x95c0242daa0b3113a3d9cA45b4E6421a10DB77b6',
        homestead: '0xB4B1767eF6f1ce65DeEBa3F0AeFa67E6CFd40639',
    },
};
exports.constants = constants;
