// @onchain-id/identity-sdk/dist/utils-module.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.contracts = exports.enums = exports.crypto = void 0;
const Claim_interface_1 = require("./claim/Claim.interface");
const Key_interface_1 = require("./identity/Key.interface");
const solidity_1 = require("@onchain-id/solidity");
const crypto = __importStar(require("./crypto-module"));
exports.crypto = crypto;
const enums = {
    ClaimScheme: Claim_interface_1.ClaimScheme,
    ClaimType: Claim_interface_1.ClaimTopic,
    KeyPurpose: Key_interface_1.KeyPurpose,
    KeyType: Key_interface_1.KeyType,
};
exports.enums = enums;
const contracts = {
    ERC734: solidity_1.interfaces.IERC734,
    ERC735: solidity_1.interfaces.IERC735,
    Identity: solidity_1.contracts.Identity,
};
exports.contracts = contracts;
__exportStar(require("./core/utils/Utils"), exports);
